exonum_derive/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+            .iter()
+            .find_map(|attr| {
+                let meta = attr.parse_meta().ok()?;
+                if meta.name() != "protobuf_convert" {
+                    return None;
+                }
+                let list = match meta {
+                    Meta::List(x) => x,
+                    _ => panic!("protobuf_convert attribute expects one argument"),
+                };
+                let name: Path = match list.nested.iter().next().expect("h") {
+                    NestedMeta::Literal(Lit::Str(lit_str)) => lit_str
+                        .parse()
+                        .expect("protobuf_convert argument should be valid type path"),
+                    _ => panic!("protobuf_convert argument should be string"),
+                };
+                Some(name)
+            }).expect("protobuf_convert attribute not set")
+    };
+    let mod_name = Ident::new(&format!("pb_convert_impl_{}", name), Span::call_site());
+    let data = match input.data {
+        Data::Struct(x) => x,
+        _ => panic!("Protobuf convert can be derived for structs only"),
+    };
+
+    let field_names = data
+        .fields
+        .iter()
+        .map(|f| f.ident.clone().unwrap())
+        .collect::<Vec<_>>();
+
+    let to_pb_fn = {
+        let names1 = field_names.clone();
+        let setters = names1
+            .iter()
+            .map(|i| Ident::new(&format!("set_{}", i), Span::call_site()));
+        let names2 = field_names.clone();
+        quote! {
+            fn to_pb(&self) -> Self::ProtoStruct {
+                let mut msg = Self::ProtoStruct::new();
+                #( msg.#setters(self.#names2.to_pb().into()); )*
+                msg
+            }
+        }
+    };
+
+    let from_pb_fn = {
+        let names1 = field_names.clone();
+        let getters = names1
+            .iter()
+            .map(|i| Ident::new(&format!("get_{}", i), Span::call_site()));
+        let names2 = field_names.clone();
+        quote! {
+            fn from_pb(pb: Self::ProtoStruct) -> Result<Self, ()> {
+              Ok(Self {
+               #( #names2: ProtobufConvert::from_pb(pb.#getters().to_owned())?, )*
+              })
+            }
+        }
+    };
+
+    let cr = if env::var("CARGO_PKG_NAME").unwrap() == "exonum" {
+        quote!(crate)
+    } else {
+        quote!(exonum)
+    };
+
+    let expanded = quote! {
+        mod #mod_name {
+            use super::*;
+
+            use std::borrow::Cow;
+
+            use #cr::messages::BinaryForm;
+            use #cr::encoding::protobuf::ProtobufConvert;
+            use #cr::storage::StorageValue;
+            use #cr::encoding;
+            use protobuf::Message;
+            use crypto::{self, CryptoHash, Hash};
+
+            impl ProtobufConvert for #name {
+                type ProtoStruct = #proto_struct_name;
+
+                #to_pb_fn
+
+                #from_pb_fn
+
+            }
+
+            impl BinaryForm for #name
+            {
```

Comment:
This lone opening brace (together the empty line above) trigger and offend my perfectionism. I guess they are not covered by rustfmt so please don't let them stand out of the code style.